ql/experimental/math/laplaceinterpolation.cpp

```diff
@@ -0,0 +1,253 @@
+                        coord_j[j] = corner_neighbour_index[j];
+                        Real weight = 0.0;
+                        for (Size i = 0; i < dim.size(); ++i) {
+                            if (i != j)
+                                weight += corner_h[i];
+                        }
+                        weight = dim.size() == 1 ? 1.0 : weight / sum_corner_h;
+                        g(count, layout_->index(coord_j)) = -weight;
+                    }
+                    g(count, count) = 1.0;
+                } else {
+                    // point with at least one dimension with non-trivial second derivative
+                    for (auto colit = rowit.begin(); colit != rowit.end(); ++colit)
+                        g(count, colit.index2()) = *colit;
+                }
+                rhs[count] = 0.0;
+                guess[count] = guessTmp;
+            } else {
+                g(count, count) = 1;
+                rhs[count] = val;
+                guess[count] = guessTmp = val;
+            }
+            ++count;
+            ++rowit;
+        }
+
+        interpolatedValues_ = BiCGstab(f_A(g), 10 * N, relTol_).solve(rhs, guess).x;
+    }
+
+    std::vector<Size>
+    LaplaceInterpolation::projectedCoordinates(const std::vector<Size>& coordinates) const {
+        std::vector<Size> tmp;
+        for (Size i = 0; i < coordinates.size(); ++i) {
+            if (coordinateIncluded_[i])
+                tmp.push_back(coordinates[i]);
+        }
+        return tmp;
+    }
+
+    std::vector<Size>
+    LaplaceInterpolation::fullCoordinates(const std::vector<Size>& projectedCoordinates) const {
+        std::vector<Size> tmp(coordinateIncluded_.size(), 0);
+        for (Size i = 0, count = 0; i < coordinateIncluded_.size(); ++i) {
+            if (coordinateIncluded_[i])
+                tmp[i] = projectedCoordinates[count++];
+        }
+        return tmp;
+    }
+
+    Real LaplaceInterpolation::operator()(const std::vector<Size>& coordinates) const {
+        QL_REQUIRE(coordinates.size() == x_.size(), "LaplaceInterpolation::operator(): expected "
+                                                        << x_.size() << " coordinates, got "
+                                                        << coordinates.size());
+        if (numberOfCoordinatesIncluded_ == 0) {
+            Real val = y_(coordinates);
+            return val == Null<Real>() ? 0.0 : val;
+        } else {
+            return interpolatedValues_[layout_->index(numberOfCoordinatesIncluded_ == x_.size() ?
+                                                          coordinates :
+                                                          projectedCoordinates(coordinates))];
+        }
+    }
+
+    template <class M>
+    void laplaceInterpolation(M& A,
+                              const std::vector<Real>& x,
+                              const std::vector<Real>& y,
+                              Real relTol) {
```

Comment:
Is this here in the cpp because we only want it to work with Matrix?  If so, we can make it a non-template function taking a Matrix.

Reply:
I agree, I'll make that a non-template function taking a QuantLib::Matrix. I am not sure what other classes M I had in mind 10 years ago...